### DISCUSSION TITLE
Difference between Loading Graph from CSV and Creating Custom Dataset.

### DISCUSSION THREAD

**Discussion by mprithiv** [pyg-team/pytorch_geometric#3866] [Q&A]
Hi,

I am trying to create an unsupervised graph embedding using heterogenous graph. I am using my own dataset that I have in a CSV format. I tried to use "Load Graph from CSV" but then when trying to create a GNN model as per the example, I am stuck with not having train_mask. Does this mean, I need to create a dataset first and then load the csv files? What is the difference between loading graph from CSV and Creating a custom dataset?

Thanks

**Comment by rusty1s** [comment]
If you do not have a dedicated train mask, you can create one after graph creation as well, e.g., via the [`torch_geometric.transforms.RandomNodeSplit`](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.RandomNodeSplit) transform.

The `torch_geometric.data.Dataset` comes in handy when working with multiple graphs, and it provides support for downloading, processing and saving data objects. In your case, there isn't a real difference.

**Reply by mprithiv** [reply]
Does the RandomNodeSplit work for HeteroData object or train mask has to be created for individual nodes? Is there a way to split on a HeteroData object. 

With the data object looking like this,

`HeteroData(
  buyerId={
    num_nodes=7813,
    x=[7813, 1152]
  },
  sellerId={
    num_nodes=884,
    x=[884, 387]
  },
  (buyerId, txn_amount, sellerId)={ edge_index=[2, 335959] },
  (buyerId, txn_date, sellerId)={ edge_indes=[2, 335959] }
)`

performing a random RandomNodeSplit on the whole HeteroData fails. The task is, creating an embedding for the buyer and seller with txn_amount and date as edges, this embedding will be used for many downstream applications. I am intended to try the GAT model under [Creating Heterogeneous graphs](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html)

**Reply by rusty1s** [reply]
I see that this might be iriritating. The `RandomNodeSplit` will only split on node types that contain a certain key (see the `key` attribute of `RandomNodeSplit`. You may try to set it to `None`.

**Reply by mprithiv** [reply]
I tried setting the key to None. Below represents the data object

`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (buyer, txn_amount, seller)={ edge_index=[2, 335959] },
  (buyer, txn_date, seller)={ edge_indes=[2, 335959] }
)
`

There is no common feature column between buyer and seller. The mapping has buyerId, sellerId, tnx_amount and txn_date. Splitting with key=none

`nodesplit = RandomNodeSplit(split='random', key=None)
nodesplit(data)`

results in
`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (buyer, txn_amount, seller)={},
  (buyer, txn_amount, seller)={ edge_index=[2, 335959] },
  (buyer, txn_date, seller)={ edge_indes=[2, 335959] },
  (seller, rev_txn_amount, buyer)={ edge_index=[2, 335959] }
)`



I don't see any split happening. Whereas, for RandomLinkSplit

`edge_split = RandomLinkSplit(
    num_val=0.05,
    num_test=0.1,
    neg_sampling_ratio=0.0,
    edge_types=[('buyer', 'txn_amount', 'seller')],
    rev_edge_types=[('seller', 'rev_txn_amount', 'buyer')],
)`

results in
Train Data
====================
`HeteroData(
  user={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (user, txn_amount, seller)={},
  (user, txn_amount, seller)={
    edge_index=[2, 285567],
    edge_label=[285567],
    edge_label_index=[2, 285567]
  },
  (user, txn_date, seller)={ edge_indes=[2, 335959] },
  (seller, rev_txn_amount, user)={ edge_index=[2, 285567] }
)`

Test Data
====================
`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (buyer, txn_amount, seller)={},
  (buyer, txn_amount, seller)={
    edge_index=[2, 302364],
    edge_label=[33595],
    edge_label_index=[2, 33595]
  },
  (buyer, txn_date, seller)={ edge_indes=[2, 335959] },
  (seller, rev_txn_amount, buyer)={ edge_index=[2, 302364] }
)`

Validation Data
====================
`HeteroData(
  buyer={
    num_nodes=7813,
    x=[7813, 1152]
  },
  seller={
    num_nodes=884,
    x=[884, 387]
  },
  (buyer, txn_amount, seller)={},
  (buyer, txn_amount, seller)={
    edge_index=[2, 285567],
    edge_label=[16797],
    edge_label_index=[2, 16797]
  },
  (buyer, txn_date, seller)={ edge_indes=[2, 335959] },
  (seller, rev_txn_amount, buyer)={ edge_index=[2, 285567] }
)`

Also, 

- I see reverse-edge being created for only one edge types after performing transform = T.ToUndirected()(data_copy). Not sure if this is normal. 
- When I do a node split/edge split on a heterogeneous graph with multiple node and edge types like aforesaid, how could I make sure the split happen across all the node/edge types?

**Reply by rusty1s** [reply]
I think you need to use `split="train_rest"` in order to allow for splits independent of any ground-truth labels.
